### PR TITLE
Let users specify custom Ask message on CLI

### PR DESCRIPTION
**Pull request by goto-bus-stop**
Opens the $EDITOR in order to add a message.

A temporary file is placed in `profiledir/ASK_MESSAGE`. If the profile is
super broken and didn't output a directory, it uses a temporary file
… somewhere (whatever the `tmp` module decides to do).

### CODE REVIEW

**Comment by AlanSl** [comment]
If something went wrong doing an ask, then you ask again, you get this:

image

Could we just import and/or delete the existing swp file if there is one?

**Comment by AlanSl** [comment]
A few other things:

1. For large profiles, you have to wait for the upload to complete before the text editor opens. Could the upload be happening while the user is typing their message, then confirm it when the message is sent?

2. Can we have a couple of blank newlines after the dummy text so the user can get to the input bit with a few down key presses (and if possible have the cursor start at that position)?

3. Sometimes it just flashes on pressing return then shows this message on typing, any idea why?

image

4. Can we explicitly say in the comment default text what the user needs to do to submit their input?

**Comment by goto-bus-stop** [comment]
we can disable the swap file entirely using `-n`, but i don't know if that behaviour is the same for different editors.

removing the swap file is hard because its location depends on the vim configuration, and is very commonly somewhere else (like in `~/.vim/swap` or `/tmp`). and if $EDITOR is not vim, it'll be totally different again.

> Sometimes it just flashes on pressing return then shows this message on typing, any idea why?

you're probably not in insert mode :thinking: 

> Can we explicitly say in the comment default text what the user needs to do to submit their input?

what sort of thing specifically? i'd be a little wary of putting a quick vim tutorial (hit `i`, type, hit Esc, type `:wq`) in there because the editor might be something else.

we could check up front if the editor is vim, and put in a little more guidance in that case.

File: bin.js [hunk 1]
```diff
@@ -626,12 +626,15 @@ async function getAskMessage (dirname) {
     # Asking for help with: ${basename}
     # Please enter your question below. There is no specific formatting, but
     # feel free to use Markdown if you need to.
+    #
+    # You may have been dropped into vim. If you're unfamiliar with it, do:
+    # - press 'i' to enter insert mode
```
Comment (AlanSl) [inline bin.js]: Can we open Vim in insert mode, e.g. `+star`? I imagine frequent Vim users would see that they're in insert mode, and everyone else could just start typing, no surprises.

If so, we could replace all this handholding with one conditional line, "When finished, press Escape then `:wq` to save and exit" if it's Vim that is being opened (I think you said we can detect this?)

**Comment by goto-bus-stop** [comment]
Tweaks:
- `getAskMessage` gets its own file
- vim starts in insert mode with the cursor at the end of the file
- VS Code forks and returns control to the CLI on start, so we can't
easily wait for it to exit; instead, we wait for user confirmation if
the editor closed within 2 seconds without updating the ASK_MESSAGE file.

I used `inquirer` for the user confirmation thing, it's a really big dependency but it's already in use via `insight` so :woman_shrugging:

File: lib/get-ask-message.js [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+  let header = dedent`
+    # Asking for help with: ${basename}
+    # Please enter your question below. There is no specific formatting, but
+    # feel free to use Markdown if you need to.
+  `
+  if (editor === 'vim') {
+    header += '\n' + dedent`
+      #
+      # You've been dropped into vim, already in 'insert mode'.
+      # After writing your message, press Escape, and then type ':wq' (no quotes) to save and exit.
```
Comment (AlanSl) [inline lib/get-ask-message.js]:
```suggestion
      # After writing your message, to save and exit, press Escape then type :wq
```
Shorter and means we don't need quotes and "no quotes"

Comment (goto-bus-stop) [reply lib/get-ask-message.js]: clever :)

**Pull request merged**